AGENTS.md

```diff
@@ -0,0 +1,187 @@
+
+## Building and Testing
+
+### Build commands
+```bash
+# Build the CVO binary
+make build                # Runs hack/build-go.sh, outputs to _output/
+
+# Build the CVO container image
+./hack/build-image.sh     # Creates local image with git version tag
+
+# Push development image to a registry
+REPO=quay.io/yourname ./hack/push-image.sh
+
+# Format code
+make format              # Run go fmt on all packages
+```
+
+### Testing commands
+```bash
+# Unit tests
+make test               # Runs gotestsum with JUnit XML output to _output/junit.xml
+
+# Single package test
+gotestsum --packages="./pkg/cvo"
+
+# Integration tests (requires KUBECONFIG with admin credentials for a disposable cluster)
+make integration-test   # Runs tests with TEST_INTEGRATION=1 against real cluster
+
+# Update test metadata (for tests in test/ directory)
+make update
+
+# Verification
+make verify             # Runs verify-yaml and verify-update
+make verify-yaml        # Validates YAML manifests
+make verify-update      # Ensures generated files are up-to-date
+```
+
+## Architecture
+
+### Core Components
+
+#### Main Operator Loop (pkg/cvo/cvo.go)
+- The `Operator` struct is the central controller
+- Watches the `ClusterVersion` resource in cluster
+- Implements a reconciliation loop that is the single writer of `ClusterVersion` status
+- Uses a work queue pattern with 15 max retries for failed operations
+- Periodically checks for updates from configured upstream server
+- Maintains conditions: `Progressing`, `Available`, `Failing`, `Upgradeable`, `RetrievedUpdates`
+
+#### SyncWorker (pkg/cvo/sync_worker.go)
+- Abstracts payload synchronization via `ConfigSyncWorker` interface
+- Manages the actual application of manifests from release payload images
+- Handles payload retrieval, verification, and application
+- Uses rate limiting and retry logic for resilient updates
+- Reports status back to main operator via `SyncWorkerStatus` channel
+
+#### Payload Management (pkg/payload/)
+- `State` enum defines update modes: `UpdatingPayload`, `ReconcilingPayload`, `InitializingPayload`, `PrecreatingPayload`
+- `UpdatingPayload`: Conservative ordering when transitioning versions, errors block dependent operators
+- `ReconcilingPayload`: Maintains current state, recreates resources without strict ordering
+- `InitializingPayload`: First-time deployment, fast progress with tolerance for transient errors
+- Release payloads stored in two directories:
+  - `manifests/`: CVO's own manifests
+  - `release-manifests/`: Manifests for other cluster operators
+- Task graph orchestrates ordered application of manifests based on dependencies
+
+#### Resource Libraries (lib/)
+- `resourcebuilder/`: Builds Kubernetes resources from manifests
+- `resourceapply/`: Applies resources with proper merge semantics for different API types
+- `resourcemerge/`: Merges resource specifications (handles fields like `hostUsers` flag)
```

Comment:
`hostUsers` seems like a weirdly-specific call-out :sweat_smile: Maybe say something more generic about merging `spec` properties, and currently needing to be taught about which specific properties need reconciling (vs. the ones we ignore and are allowed to drift around without active CVO management)?

Reply:
However, I am learning all this fun stuff, I am happy to change anything.